### PR TITLE
nixos/vsftpd: add userlist type

### PR DESCRIPTION
**Pull request by Scriptkiddi**
###### Motivation for this change
#76184

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by yorickvP** [comment]
This individual change is fine, but userlistFile should be filled in from this option as something stronger than a default, to prevent careless overriding. Something like 
`services.vsftpd.userlistFile = lib.mkIf (cfg.userlist != []) (pkgs.writeText "userlist" (concatMapStrings (x: "${x}\n") cfg.userlist);`?

**Comment by Scriptkiddi** [comment]
@yorickvP like this?

File: nixos/modules/services/networking/vsftpd.nix [hunk 1]
```diff
@@ -300,10 +304,13 @@ in
 
     users.groups.ftp.gid = config.ids.gids.ftp;
 
+
```
Comment (cole-h) [inline nixos/modules/services/networking/vsftpd.nix:307]:
```suggestion
```

File: nixos/modules/services/networking/vsftpd.nix [hunk 2]
```diff
@@ -300,10 +304,13 @@ in
...
     # If you really have to access root via FTP use mkOverride or userlistDeny
     # = false and whitelist root
     services.vsftpd.userlist = if cfg.userlistDeny then ["root"] else [];
 
+    services.vsftpd.userlistFile = lib.mkIf (cfg.userlist != []) (pkgs.writeText "userlist" (concatMapStrings (x: "${x}\n") cfg.userlist));
```
Comment (cole-h) [inline nixos/modules/services/networking/vsftpd.nix:312]: Should these be combined?

```
services.vsftpd = {
  userlist = ...;
  userlistFile = ...;
};

**Review by yorickvP — APPROVED** [review]

**Review by yorickvP — CHANGES_REQUESTED** [review]
I believe the assertion won't work, since we override userListFile in this package if someone sets userList. However, the check will still happen without the assertion since you wouldn't be able to merge in another userListFile.

**Pull request closed without merging**